src/lib.rs

```diff
@@ -426,3 +460,111 @@ fn get_repo_teams(bootstrap: &Bootstrap, repo: &str) -> Result<HashSet<Team>, St
         None,
     )
 }
+
+/// Get the visibility of a repository, given its name
+fn get_repo_visibility(bootstrap: &Bootstrap, repo: &str) -> Result<String, String> {
+    let res = make_github_request(
+        &bootstrap.token,
+        &format!("/repos/{}/{repo}", bootstrap.org),
+        3,
+        None,
+    )?;
+    res.get("visibility")
+        .and_then(|v| v.as_str())
+        .and_then(|v| Some(v.to_string()))
+        .ok_or("Missing visibility in the response from GitHub".to_string())
+}
+
+#[derive(serde::Serialize)]
+pub struct GraphQLQuery {
+    query: String,
+    variables: HashMap<String, String>,
+}
+
+/// Make a GraphQL query on GitHub
+fn make_graphql_query(
+    gh_token: &str,
+    query: GraphQLQuery,
+    retries: u8,
+) -> Result<serde_json::Value, String> {
+    let query = serde_json::to_string(&query).unwrap();
+
+    let mut tries = 0;
+    loop {
+        tries += 1;
+        let response = reqwest::blocking::Client::new()
+            .post(GRAPHQL_URL)
+            .header("User-Agent", "GitHub EC Audit")
+            .header("Accept", "application/vnd.github+json")
+            .header("X-GitHub-Api-Version", "2022-11-28")
+            .header("Authorization", format!("Bearer {}", gh_token))
+            .body(query.clone())
+            .send()
+            .map(|response| response.text());
+
+        // Handle communication issues with GitHub
+        let content = match response {
+            Ok(Ok(content)) => content,
+            Ok(Err(e)) => {
+                if tries >= retries {
+                    println!("{}", "Retries exhausted".red());
+                    return Err(e.to_string());
+                }
+
+                println!(
+                    "{}: {}",
+                    "Going to retry because couldn't read response from GitHub:".yellow(),
+                    e.to_string().red()
+                );
+
+                continue;
+            }
+            Err(e) => {
+                if tries >= retries {
+                    println!("{}", "Retries exhausted".red());
+                    return Err(e.to_string());
+                }
+
+                println!(
+                    "{}: {}",
+                    "Going to retry because couldn't make request to GitHub:".yellow(),
+                    e.to_string().red()
+                );
+
+                continue;
+            }
+        };
+
+        let value = serde_json::from_str::<serde_json::Value>(&content)
+            .map_err(|e| format!("Could not deserialize GitHub's response. Error: {e}"))?;
+
+        // break the loop and return the value we got
+        return Ok(value);
+    }
+}
+
+/// Get the email address associated to a username, if available, through the configured SAML IdP.
+fn email_from_gh_username(bootstrap: &Bootstrap, user: impl Display) -> Option<String> {
```

Comment:
Should be named more relative to getting an SSO email. Otherwise this will get confusing if we ever need to get emails of ECs for example.